### DISCUSSION TITLE
eth_signTypedData_v4 not available

### DISCUSSION THREAD

**Discussion by TrueCarry** [WalletConnect/walletconnect-monorepo#2880] [Developer Support] [answered]
I've used `eth_signTypedData_v4` with simple ethereum.request call. 
I've tried to call it in v2:
```
await ethereumClient.wagmi.publicClients.get(97).request({
        method: 'eth_signTypedData_v4',
        params: [
          address,
          jsoni,
        ],
      })
```

But viem returns error: 
```
Details: the method eth_signTypedData_v4 does not exist/is not available
Version: viem@1.2.5
```

**Comment by jxom** [comment]
You'll need to add `eth_signTypedData_v4` as an optional method in WalletConnect's `EthereumProvider`

**Reply by TrueCarry** [reply]
Sorry, but how do I do that in Web3Modal?

**Reply by finessevanes** [reply]
@TrueCarry what Web3Modal package are you using?

**Reply by TrueCarry** [reply]
```
let publicClient
let wagmiConfig
let ethereumClient
let web3Modal

  const { EthereumClient, w3mConnectors, w3mProvider } = require('@web3modal/ethereum')
  const { Web3Modal } = require('@web3modal/html')
  const { configureChains, createConfig } = require('@wagmi/core')
  const { bsc, bscTestnet } = require('@wagmi/core/chains')

  const chains = [{
    ...bsc,

  }, bscTestnet]
  const projectId = 'XXX'

  const config = configureChains(chains, [w3mProvider({ projectId })])
  publicClient = config.publicClient

  wagmiConfig = createConfig({
    autoConnect: true,
    connectors: w3mConnectors({ projectId, chains }),
    publicClient,
  })
  ethereumClient = new EthereumClient(wagmiConfig, chains)
  web3Modal = new Web3Modal({ projectId }, ethereumClient)
```

**Reply by TrueCarry** [reply]
@finessevanes can you help, please?

**Comment by jeremy-kunzhou** [comment]
Hi @jxom ,
We are facing the same problem when using WC_V2 with metamask using eth_signTypedData_v4
We are using @walletconnect/modal-react-native 1.0.0-rc.3. 
We add eth_signTypedData_v4 in methods list in namespance, but not working. After call, return internal error.

Many thanks.

**Comment by ndr** [comment]
The same problem:

1. Initialization:
``` 
import {EthereumProvider} from "@walletconnect/ethereum-provider";
import {providers} from "ethers";

const ethereumProvider = await EthereumProvider.init({
    projectId: "XXXXXX",
    chains: [43114],
    showQrModal: true,
    optionalMethods: [
        "eth_signTypedData",
        "eth_signTypedData_v4",
        "eth_sign",
    ],
});
if (ethereumProvider.accounts > 0) {
    const accounts = await ethereumProvider.enable();
    const provider = new providers.Web3Provider(walletConnectProvider);
    const signer = provider.getSigner();
    commit("setWalletProviderName", "WalletConnect");
    commit("setProvider", provider);
    commit("setAccount", accounts[0]);
    commit("setSigner", signer);
    commit("setWalletConnection", true);
    return signer;
}
```

2. Trying to signTypedData:
```
// domain {
//     "name": "BentoBox V1",
//     "chainId": 43114,
//     "verifyingContract": "0xXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
// }


// types {
//     "SetMasterContractApproval": [
//         {
//             "name": "warning",
//             "type": "string"
//         },
//         {
//             "name": "user",
//             "type": "address"
//         },
//         {
//             "name": "masterContract",
//             "type": "address"
//         },
//         {
//             "name": "approved",
//             "type": "bool"
//         },
//         {
//             "name": "nonce",
//             "type": "uint256"
//         }
//     ]
// }

// value {
//     "warning": "Give FULL access to funds in (and approved to) BentoBox?",
//     "user": "0xXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
//     "masterContract": "0xXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
//     "approved": true,
//     "nonce": {
//         "type": "BigNumber",
//         "hex": "0x00"
//     }
// }
this.signer
    ._signTypedData(domain, types, value)
```
And keep getting this error: {context: 'client'} 'Internal error'
And then "Uncaught (in promise) Error: No matching key. history: 1688378369243776"

ethers: v5.7.2
@walletconnect/provider v2.8.6


Any idea, please

**Reply by Abstrucked** [reply]
I'm getting the exact same..

**Comment by TrueCarry** [comment]
I've added mainnet to networks list. It fixed sign_typeddata_v4, but I'm not sure if it breaks something yet. @ndr, @jeremy-kunzhou

**Reply by ndr** [reply]
@TrueCarry I tried your workaround, but no success. What versions of packages are you using?

**Reply by TrueCarry** [reply]
Latest at the moment of the first post. Too much packages, not sure what version is relevant.

**Reply by paintoshi** [reply]
Interestingly, after much investigation and no success, this was the one fix that solved it for me. I was only using Fantom as the configured network in the Web3Modal. I don't need mainnet but I can't see how it would do any harm sitting in there. So basically just:

```
import { mainnet, fantom } from 'wagmi/chains'
additionalChains = [mainnet, fantom]
const { publicClient } = configureChains(additionalChains, [w3mProvider({ projectId })])
const wagmiConfig = createConfig({
  autoConnect: true,
  connectors: w3mConnectors({ projectId, chains: additionalChains }),
  publicClient,
})

const ethereumClient = new EthereumClient(wagmiConfig, additionalChains)
```

Using:

```
"@web3modal/ethereum": "^2.7.1",
"@web3modal/react": "^2.7.1",
"viem": "^1.4.2",
"wagmi": "^1.3.9",
```

**Reply by zlace0x** [reply]
Adding mainnet worked for me too!

**Reply by SergioArrighi** [reply]
It's incredible but for some reason this worked for me too!
Maybe it is a wagmi issue then?
Thanks and best regards

**Reply by SergioArrighi** [reply]
Hello guys,

so I figured this out.
I found this issue in wagmi repo
https://github.com/wagmi-dev/wagmi/issues/3012
Basically the only thing that matters is that you add at least two chains in the connectors configuration.
The best way to work around this issue is to add the same network twice if you, like me, are only using a single network.
Thanks and best regards.

**Comment by TrueCarry** [comment, marked as answer]
After many attempts to make it work, I just switched to EIP-4361

**Reply by ikwuoz** [reply]
Hi can you give a breakdown of how you implemented the switch? @TrueCarry

**Reply by ComradeAERGO** [reply]
@TrueCarry I'm seconding this request, this would be extremely helpful.

**Reply by TrueCarry** [reply]
I don't know what to explain. It's covered in guides. Basically take any lib that supports generating of EIP-4361 messages, create your message, send to wallet. Conceptually not much changed, just different signing method and payload generator

**Comment by melnychuk-anton** [comment]
wow walletconnect, what a solution.. people have to rewrite their applications to work with your new improved version, which is totally broken and doesn't even support running your own instance, offering to give all the details of your project to get a project id to be able to use it instead (trully decentralized). you can write all over your documentation that you do not recommend to stay on v1 but that's the only solution I see for now.

**Comment by hironate** [comment]
Is there any solution to this, I am getting this error when calling the seaport signOrder function. i am using Ethers adaptor by Wagmi to get a signer, as seaport only supports Ethers.js signer. i don't know why with walletConnect it is calling this RPC eth_signTypedData_v4 with metamask works fine.

**Comment by drinkius** [comment]
Same question here, would like to get an update
